# Prediction of a potential car purchasor

Research objective: In this projrct we shall try to find out an appropriate model for predicting                           wheather a person will buy a car or not based on different attributrs like                             age,annual salary, gender etc


In [1]:
#lets import the data set
import pandas as pd
df=pd.read_csv("C:/Users/ronit/OneDrive/Desktop/data science/project data/data for classification/car_data.csv")
df

User ID  Gender  Age  AnnualSalary  Purchased
0        385    Male   35         20000          0
1        681    Male   40         43500          0
2        353    Male   49         74000          0
3        895    Male   40        107500          1
4        661    Male   25         79000          0
..       ...     ...  ...           ...        ...
995      863    Male   38         59000          0
996      800  Female   47         23500          0
997      407  Female   28        138500          1
998      299  Female   48        134000          1
999      687  Female   44         73500          0

[1000 rows x 5 columns]

Methodology: After cleaning and preprocessing the data we shall run a logit model,
             a decisoion tree model and a K-NN model for prediction of the potential
             car purchasor. We shall check the accuracy score of every model. Based on 
             the accuracy score we shall rank the model. Ultimately we shall use the 
             model which has highest accuracy score for prediction of a potential car
             purchasor

In [2]:
df['Gender'].unique()

array(['Male', 'Female'], dtype=object)

In [3]:
df['Gender'].value_counts()

Female    516
Male      484
Name: Gender, dtype: int64

In [4]:
df.describe()

User ID          Age   AnnualSalary    Purchased
count  1000.000000  1000.000000    1000.000000  1000.000000
mean    500.500000    40.106000   72689.000000     0.402000
std     288.819436    10.707073   34488.341867     0.490547
min       1.000000    18.000000   15000.000000     0.000000
25%     250.750000    32.000000   46375.000000     0.000000
50%     500.500000    40.000000   72000.000000     0.000000
75%     750.250000    48.000000   90000.000000     1.000000
max    1000.000000    63.000000  152500.000000     1.000000

In [5]:
df1=pd.get_dummies(df)
df1

User ID  Age  AnnualSalary  Purchased  Gender_Female  Gender_Male
0        385   35         20000          0              0            1
1        681   40         43500          0              0            1
2        353   49         74000          0              0            1
3        895   40        107500          1              0            1
4        661   25         79000          0              0            1
..       ...  ...           ...        ...            ...          ...
995      863   38         59000          0              0            1
996      800   47         23500          0              1            0
997      407   28        138500          1              1            0
998      299   48        134000          1              1            0
999      687   44         73500          0              1            0

[1000 rows x 6 columns]

In [6]:
import statsmodels.api as sm

In [7]:
result=sm.Logit(df1['Purchased'],sm.add_constant(df1[['AnnualSalary','Age','Gender_Male']])).fit()

Optimization terminated successfully.
         Current function value: 0.371479
         Iterations 7


In [8]:
print(result.summary())

                           Logit Regression Results                           
Dep. Variable:              Purchased   No. Observations:                 1000
Model:                          Logit   Df Residuals:                      996
Method:                           MLE   Df Model:                            3
Date:                Thu, 18 Aug 2022   Pseudo R-squ.:                  0.4487
Time:                        08:42:01   Log-Likelihood:                -371.48
converged:                       True   LL-Null:                       -673.81
Covariance Type:            nonrobust   LLR p-value:                9.794e-131
                   coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
const          -12.1788      0.805    -15.131      0.000     -13.756     -10.601
AnnualSalary   3.37e-05   3.23e-06     10.426      0.000    2.74e-05       4e-05
Age              0.2195      0.015     14.47

# Let's go for the accuracy checking of the logit model

In [9]:
from sklearn import linear_model
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

In [10]:
x=df1.drop(['Purchased','User ID'],axis='columns')
y=df1.drop(['User ID','Age','AnnualSalary','Gender_Female','Gender_Male'],axis='columns')

In [11]:
x

Age  AnnualSalary  Gender_Female  Gender_Male
0     35         20000              0            1
1     40         43500              0            1
2     49         74000              0            1
3     40        107500              0            1
4     25         79000              0            1
..   ...           ...            ...          ...
995   38         59000              0            1
996   47         23500              1            0
997   28        138500              1            0
998   48        134000              1            0
999   44         73500              1            0

[1000 rows x 4 columns]

In [12]:
x.columns

Index(['Age', 'AnnualSalary', 'Gender_Female', 'Gender_Male'], dtype='object')

In [13]:
y

Purchased
0            0
1            0
2            0
3            1
4            0
..         ...
995          0
996          0
997          1
998          1
999          0

[1000 rows x 1 columns]

In [14]:
y.columns

Index(['Purchased'], dtype='object')

In [15]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.25
                                              )

In [16]:
model1=LogisticRegression()

In [17]:
result=model1.fit(x_train,y_train)

C:\Users\ronit\anaconda3\lib\site-packages\sklearn\utils\validation.py:72: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


In [18]:
y_predict=model1.predict(x_test)
y_predict

array([0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1,
       1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1,
       0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0,
       0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1,
       0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1,
       0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 1, 0, 1, 0, 0, 1, 0, 0,
       0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1,
       0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 0, 0, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0,
       1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 1,
       0, 0, 1, 1, 0, 0, 0, 0], dtype=int64)

In [19]:
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

In [20]:
accuracy_score(y_test,y_predict)

0.792

In [21]:
tn,fp,fn,tp=confusion_matrix(y_test,y_predict).ravel()
print('true negative:',tn)
print('false positive:',fp)
print('false negative:',fn)
print('true positiv:',tp)

true negative: 125
false positive: 25
false negative: 27
true positiv: 73


In [22]:
#generating classification report
from sklearn.metrics import classification_report
print(classification_report(y_test,y_predict))

              precision    recall  f1-score   support

           0       0.82      0.83      0.83       150
           1       0.74      0.73      0.74       100

    accuracy                           0.79       250
   macro avg       0.78      0.78      0.78       250
weighted avg       0.79      0.79      0.79       250



# Now we shall go to build a decision tree

In [23]:
from sklearn import tree

In [24]:
mode2=tree.DecisionTreeClassifier()

In [25]:
mode2.fit(x_train,y_train)

DecisionTreeClassifier()

In [26]:
mode2.score(x_test,y_test)

0.9

# Lets use a KNN model

In [27]:
from sklearn import neighbors

In [28]:
model3=neighbors.KNeighborsClassifier(n_neighbors=5)

In [29]:
model3.fit(x_test,y_test)

<ipython-input-29-0788b5adcfda>:1: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  model3.fit(x_test,y_test)


KNeighborsClassifier()

In [30]:
model3.score(x_test,y_test)

0.88

conclusion: we have seen that our logit model has accuracy score 79% , K-NN model(with 5 nearest neighbor) has 88% accuracy and our decision tree has accuracy score 90%. So we shall use the Decision tree classification model to predict the potential buyer given age, annual salary, and gender of a person.